consensus/src/block.rs

```diff
@@ -178,17 +278,185 @@ where
                 } => {
                     verify_main_chain_block(block, prepared_txs, context_svc, tx_verifier_svc).await
                 }
+                VerifyBlockRequest::MainChainBatchPrepareBlocks { blocks } => {
+                    batch_prepare_main_chain_block(blocks, context_svc).await
+                }
+                VerifyBlockRequest::MainChainPrepped { block, txs } => {
+                    verify_prepped_main_chain_block(block, txs, context_svc, tx_verifier_svc, None)
+                        .await
+                }
             }
         }
         .boxed()
     }
 }
 
+/// Batch prepares a list of blocks for verification.
+#[instrument(level="debug", name="batch_prep_blocks" skip_all, fields(amt=blocks.len()))]
+async fn batch_prepare_main_chain_block<C>(
+    blocks: Vec<(Block, Vec<Transaction>)>,
+    mut context_svc: C,
+) -> Result<VerifyBlockResponse, ExtendedConsensusError>
+where
+    C: Service<
+            BlockChainContextRequest,
+            Response = BlockChainContextResponse,
+            Error = tower::BoxError,
+        > + Send
+        + 'static,
+    C::Future: Send + 'static,
+{
+    let (blocks, txs): (Vec<_>, Vec<_>) = blocks.into_iter().unzip();
+
+    tracing::debug!("Calculating block hashes.");
+    let blocks: Vec<PrePreparedBlockExPOW> = rayon_spawn_async(|| {
+        blocks
+            .into_iter()
+            .map(PrePreparedBlockExPOW::new)
+            .collect::<Result<Vec<_>, _>>()
+    })
+    .await?;
+
+    // A Vec of (timestamp, HF) for each block to calculate the expected difficulty for each block.
+    let mut timestamps_hfs = Vec::with_capacity(blocks.len());
+    let mut new_rx_vm = None;
+
+    tracing::debug!("Checking blocks follow each other.");
+
+    // For every block make sure they have the correct height and previous ID
+    for window in blocks.windows(2) {
+        if window[0].block_hash != window[1].block.header.previous
+            || window[0].height != window[1].height - 1
+        {
+            tracing::debug!("Blocks do not follow each other, verification failed.");
+            Err(ConsensusError::Block(BlockError::PreviousIDIncorrect))?;
+        }
+
+        // Cache any potential RX VM seeds as we may need them for future blocks in the batch.
+        if is_randomx_seed_height(window[0].height) {
+            new_rx_vm = Some((window[0].height, window[0].block_hash));
+        }
+
+        timestamps_hfs.push((window[0].block.header.timestamp, window[0].hf_version))
+    }
```

Comment:
Only need to `[]` once.
```suggestion
    for window in blocks.windows(2) {
        let block_1 = &window[0];
        let block_2 = &window[1];
        let block_1_hash = block_1.block_hash;
        let block_1_height = block_1.height;
        let block_2_height = block_2.height;

        if block_1_hash != block_2.block.header.previous || block_1_height != block_2_height - 1 {
            tracing::debug!("Blocks do not follow each other, verification failed.");
            Err(ConsensusError::Block(BlockError::PreviousIDIncorrect))?;
        }

        // Cache any potential RX VM seeds as we may need them for future blocks in the batch.
        if is_randomx_seed_height(block_1_height) {
            new_rx_vm = Some((block_1_height, block_1_hash));
        }

        timestamps_hfs.push((block_1.block.header.timestamp, block_1.hf_version))
    }
```

Reply:
I really think the compiler would take care of this + I'd rather the readability

Reply:
It's less about optimization and more about identifier safety, e.g. the following requires 1 ident modification when refactoring:
```rust
let x = &y[0];
let _ = x.a;
let _ = x.b;
let _ = x.c;
let _ = x.d;
```
but this requires 5 line changes:
```rust
let _ = &y[0];
let _ = &y[0].a;
let _ = &y[0].b;
let _ = &y[0].c;
let _ = &y[0].d;
```
It's harder to mess up when making changes in the 1st example.

Reply:
I prefer the original syntax, but your explanation make sense. I could say its *skill issue* from the dev to mess up this but its not really serious from a codebase perspective. Therefore I agree with your suggestion

Reply:
I don't know I'm still pretty skeptical, I partially did the change though